### ISSUE TITLE
Memory-efficient attention is not selected if inputs's ndim != 4

### ISSUE THREAD

**Issue by gau-nernst** [pytorch/pytorch#144960]
### 🐛 Describe the bug

```python
from contextlib import nullcontext

import torch
import torch.nn.functional as F
from torch.nn.attention import SDPBackend, sdpa_kernel

# shape used in FLUX diffusion model
seq_len = 128 * 128
head_dim = 512

# ctx = sdpa_kernel([SDPBackend.EFFICIENT_ATTENTION])
ctx = nullcontext()

torch.cuda.reset_peak_memory_stats()
shape = (1, 1, 1, seq_len, head_dim)
q, k, v = [torch.randn(shape, dtype=torch.bfloat16, device="cuda") for _ in range(3)]
with ctx:
    F.scaled_dot_product_attention(q, k, v)
print(f"{shape}: {torch.cuda.max_memory_allocated() / 1e9:.2f} GB")

torch.cuda.reset_peak_memory_stats()
shape = (1, 1, seq_len, head_dim)
q, k, v = [torch.randn(shape, dtype=torch.bfloat16, device="cuda") for _ in range(3)]
with ctx:
    F.scaled_dot_product_attention(q, k, v)
print(f"{shape}: {torch.cuda.max_memory_allocated() / 1e9:.2f} GB")

torch.cuda.reset_peak_memory_stats()
shape = (1, seq_len, head_dim)
q, k, v = [torch.randn(shape, dtype=torch.bfloat16, device="cuda") for _ in range(3)]
with ctx:
    F.scaled_dot_product_attention(q, k, v)
print(f"{shape}: {torch.cuda.max_memory_allocated() / 1e9:.2f} GB")

torch.cuda.reset_peak_memory_stats()
shape = (seq_len, head_dim)
q, k, v = [torch.randn(shape, dtype=torch.bfloat16, device="cuda") for _ in range(3)]
with ctx:
    F.scaled_dot_product_attention(q, k, v)
print(f"{shape}: {torch.cuda.max_memory_allocated() / 1e9:.2f} GB")
```
```
(1, 1, 1, 16384, 512): 2.61 GB
(1, 1, 16384, 512): 0.11 GB
(1, 16384, 512): 2.61 GB
(16384, 512): 2.61 GB
```

If I use `ctx = sdpa_kernel([SDPBackend.EFFICIENT_ATTENTION])`, only the ones where `ndim == 4` will not error out.

**Expected behavior**: memory-efficient attention should be selected.

Possibly related: #127523 (but I don't use attention mask here)

cc: @drisspg 

### Versions

torch==2.7.0.dev20250105+cu126

**Comment by malfet** [comment]
@drisspg this is an sdpa issue, isn't it?
